docs/synapse/userguides/storm_adv_functions.rstorm

```diff
@@ -0,0 +1,708 @@
+++++++++++++++++++
+
+All functions are **invoked** using the function name preceded by the dollar sign ( ``$`` ), and by
+passing any required parameters to the function.
+
+- ``$myFunction()``
+- ``$myFunction(foo)``
+- ``$myFunction($foo)``
+- ``$myFunction(bar,baz=wheee)``
+
+Parameters can be passed as literals, variables, or keyword=$valu arguments. For example, given a
+function that takes an organization name as input:
+
+::
+  
+  function $myFunction(orgname) { <do stuff> }
+
+The name can be passed directly:
+
+::
+  
+  $myFunction("The Vertex Project")
+
+...as a variable:
+
+::
+  
+  $name="The Vertex Project"
+  $myFunction($name)
+
+...or as a keyword/value pair:
+
+::
+  
+  $myFunction(orgname="The Vertex Project")
+
+::
+  
+  $name="The Vertex Project"
+  $myFunction(orgname=$name)
+
+
+Operating on Nodes
+******************
+
+Functions do not inherit or operate on the invoking Storm pipeline by default. If you want a function
+to operate on nodes in a pipeline, you can invoke the function by passing a property (or properties)
+of the node as input to the function.
+
+For example, if your invoking pipeline consists of a set of ``inet:ipv4`` nodes, a function can take the
+``:asn`` property as input:
+
+::
+  
+  $myFunction(:asn)
+
+Or:
+
+::
+  
+  $asn=:asn
+  $myFunction($asn)
+
+Alternatively, if a function yields the node(s) you want to operate on, that function can be used as input
+to a second function. A simple example:
+
+**Function 1 (node yielder):**
+
+::
+  
+  function getIPs() {
+      //Lift 10 IPv4 addresses
+      inet:ipv4 | limit 10
+  }
+
+**Function 2 (callable function):**
+
+::
+  
+  //Takes a generator object as input
+  function counter(genr) {
+      
+      //Yield the generator content into the function pipeline
+      yield $genr
+      
+      //Print the human-readable representation of each node
+      $lib.print($node.repr())
+      
+      //Return the output of the function
+      fini { return() }
+  }
+
+**Function 2 invoked with Function 1 as input:**
+
+::
+  
+  $counter($getIPs())
+
+.. _storm-func-runtsafe
+
+Runtsafe vs. Non-Runtsafe Functions
+***********************************
+
+Just as variables may be runtime-safe (`runtsafe`_) or non-runtime-safe (`non-runtsafe`_), functions can be
+invoked in a runtsafe manner (or not) based on the parameters passed to the function.
+
+If a function is invoked with a runtsafe (typically static) value, the function is considered runtsafe.
+Similarly, if a function is invoked with a per-node, non-runtsafe value or values, the function is considered
+non-runtsafe. In the example above where the function takes an AS number as input, the function is **non-runtsafe;**
+the function takes the ``inet:ipv4:asn`` value of each node, and the function is invoked **once for each node.**
+
+If the parameter were a static AS number, the function would be invoked in a **runtsafe** manner, and the
+function would be invoked **only once** (not per-node):
```

Comment:
A function invoked in a runtsafe manner will still be invoked per-node if there are nodes in the pipeline (it will just do the same thing each time)

Reply:
revised with many more words (hopefully good ones)